__tests__/integration/melt-tokens.test.js

```diff
@@ -0,0 +1,342 @@
+
+  it('should not melt with insuficcient tokens', async done => {
+    const response = await TestUtils.request
+      .post('/wallet/melt-tokens')
+      .send({
+        token: tokenA.uid,
+        address: await wallet1.getAddressAt(1),
+        amount: 1000
+      })
+      .set({ 'x-wallet-id': wallet1.walletId });
+
+    expect(response.status).toBe(200);
+    expect(response.body.success).toBe(false);
+    expect(response.body.error).toContain('enough inputs to melt');
+    done();
+  });
+
+  // Success
+
+  it('should melt with address and change address', async done => {
+    const response = await TestUtils.request
+      .post('/wallet/melt-tokens')
+      .send({
+        token: tokenA.uid,
+        amount: 300,
+        deposit_address: await wallet1.getAddressAt(3),
+        change_address: await wallet1.getAddressAt(4),
+      })
+      .set({ 'x-wallet-id': wallet1.walletId });
+
+    expect(response.body.success).toBe(true);
+
+    await TestUtils.pauseForWsUpdate();
+
+    const addr3htr = await wallet1.getAddressInfo(3);
+    const addr3tka = await wallet1.getAddressInfo(3, tokenA.uid);
+    expect(addr3htr.total_amount_available).toBe(3);
+    expect(addr3tka.total_amount_available).toBe(0);
+
+    const addr4htr = await wallet1.getAddressInfo(4);
+    const addr4tka = await wallet1.getAddressInfo(4, tokenA.uid);
+    expect(addr4htr.total_amount_available).toBe(0);
+    expect(addr4tka.total_amount_available).toBe(500);
+
+    const balance1htr = await wallet1.getBalance();
+    const balance1tka = await wallet1.getBalance(tokenA.uid);
+    expect(balance1htr.available).toBe(2 + 3);
+    expect(balance1tka.available).toBe(800 - 300);
+
+    done();
+  });
+
+  it('should melt with deposit address only', async done => {
+    const response = await TestUtils.request
+      .post('/wallet/melt-tokens')
+      .send({
+        token: tokenA.uid,
+        amount: 100,
+        deposit_address: await wallet1.getAddressAt(5),
+      })
+      .set({ 'x-wallet-id': wallet1.walletId });
+
+    expect(response.body.success).toBe(true);
+
+    await TestUtils.pauseForWsUpdate();
+
+    const addr5htr = await wallet1.getAddressInfo(5);
+    const addr5tka = await wallet1.getAddressInfo(5, tokenA.uid);
+    expect(addr5htr.total_amount_available).toBe(1);
+    expect(addr5tka.total_amount_available).toBe(0);
+
+    const balance1 = await wallet1.getBalance(tokenA.uid);
+    expect(balance1.available).toBe(500 - 100);
+
+    const balance1htr = await wallet1.getBalance();
+    const balance1tka = await wallet1.getBalance(tokenA.uid);
+    expect(balance1htr.available).toBe(2 + 3 + 1);
+    expect(balance1tka.available).toBe(800 - 300 - 100);
+
+    done();
+  });
+
+  it('should melt with change address only', async done => {
+    const response = await TestUtils.request
+      .post('/wallet/melt-tokens')
+      .send({
+        token: tokenA.uid,
+        amount: 100,
+        change_address: await wallet1.getAddressAt(7),
```

Comment:
I don't like the fact that we need to be careful with the tests order and the address indexes we are using

Not sure what is the best way to fix it tho

What do you think about having a global counter on this file, increasing it on each test, so you can do something like `wallet1.getAddressAt(addressCounter.getNextValue())`?

```javascript
const addressCounter = {
  value: 0,
  getNextValue: function() {
    this.value = this.value + 1;
    return this.value;
  },
};
```

Reply:
I was thinking on this direction when decided to implement some caches within the `wallet` class. I didn't go very far in this direction because increasing complexity on the test code could lead to bugs on the tests themselves.

I see two ways to lessen this problem:
### Declare clear variables containing the addresses and their purposes
For example, on the closure of `describe('melt tokens')`:
```js
const wallet1Addresses = {
      initialFunds: {
        index: 0,
        address: 'WSA9tePoHSvJ2ktVGCnVfefKxTwQJizAis'
      },
      basicDestination: {
        index: 1,
        address: 'WYE86DaZcdhibA3XBXnqaXGNsGK2T4Va9W'
      },
      meltDeposit1: {
        index: 2,
        address: 'WhLfSp44WHr5m1GML6BdtpkeyDqjY6byid'
      },
      meltChange1: {
        index: 3,
        address: 'WUnTgUFXeFksQUJEfaN1c3NdyErez3mvrw'
      }
    };
```

### Making actual independent tests
Each test could make its own funding transaction using `wallet.injectFunds()` , select addresses using `wallet.getNextAddress()` ( I could easily implement a way to "mark as used" ) store the starting balance on the relevant addresses and then execute the tests. Assertions would be made on values relative to the starting situation.

They would be more verbose, a bit slower to run, but would have the advantage to not depend on each other so much and ease maintainability.

What do you think of this tradeoff? Could we try one of these solutions on the next batch of tests, maybe on `filter-utxos` and `consolidate-utxos`?